handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -313,31 +333,58 @@ public void operationComplete(Future<SslContext> future) throws Exception {
     }
 
     /**
-     * The default implementation will simply call {@link AsyncMapping#map(Object, Promise)} but
-     * users can override this method to implement custom behavior.
+     * The default implementation will simply call {@link SslContextMapping#map(String, Promise)} but users can
+     * override this method to implement custom behavior.
      *
-     * @see AsyncMapping#map(Object, Promise)
+     * NOTE: This method returns a {@code Future<?>} to accommodate the old and new way of doing things without
+     * braking the API. Users who override this method are advised to change their return value from
+     * {@code Future<SslContext>} to {@code Future<Supplier<? extends SslContext>>}.
+     *
+     * @see SslContextMapping#map(String, Promise)
+     * @deprecated the return value will at some point change from {@code Future<SslContext>} to
```

Comment:
Good point. Let me clarify an approach:
1. `select` should be calling `lookupSupplier`
2. `lookupSupplier` should call `lookup` (by default)
3. `lookup` should call the `mapping` function.

This should preserve backwards compatibility by default. The only exception is still the type change of `mapping`.

```diff
diff --git a/handler/src/main/java/io/netty/handler/ssl/SniHandler.java b/handler/src/main/java/io/netty/handler/ssl/SniHandler.java
index ec17e0e..201b25f 100644
--- a/handler/src/main/java/io/netty/handler/ssl/SniHandler.java
+++ b/handler/src/main/java/io/netty/handler/ssl/SniHandler.java
@@ -15,11 +15,6 @@
  */
 package io.netty.handler.ssl;
 
-import java.net.IDN;
-import java.net.SocketAddress;
-import java.util.List;
-import java.util.Locale;
-
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufUtil;
 import io.netty.channel.ChannelHandlerContext;
@@ -35,6 +30,7 @@ import io.netty.util.ReferenceCountUtil;
 import io.netty.util.Supplier;
 import io.netty.util.concurrent.Future;
 import io.netty.util.concurrent.FutureListener;
+import io.netty.util.concurrent.GenericFutureListener;
 import io.netty.util.concurrent.ImmediateEventExecutor;
 import io.netty.util.concurrent.Promise;
 import io.netty.util.internal.ObjectUtil;
@@ -42,6 +38,11 @@ import io.netty.util.internal.PlatformDependent;
 import io.netty.util.internal.logging.InternalLogger;
 import io.netty.util.internal.logging.InternalLoggerFactory;
 
+import java.net.IDN;
+import java.net.SocketAddress;
+import java.util.List;
+import java.util.Locale;
+
 /**
  * <p>Enables <a href="https://tools.ietf.org/html/rfc3546#section-3.1">SNI
  * (Server Name Indication)</a> extension for server side SSL. For clients
@@ -297,23 +298,23 @@ public class SniHandler extends ByteToMessageDecoder implements ChannelOutboundH
     }
 
     private void select(final ChannelHandlerContext ctx, final String hostname) throws Exception {
-        Future<?> future = lookup(ctx, hostname);
+        Future<Supplier<? extends SslContext>> future = lookupSupplier(ctx, hostname);
         if (future.isDone()) {
             if (future.isSuccess()) {
-                onSslContextOrSupplier(ctx, hostname, future.getNow());
+                onSslContextSupplier(ctx, hostname, future.getNow());
             } else {
                 throw new DecoderException("failed to get the SslContext for " + hostname, future.cause());
             }
         } else {
             suppressRead = true;
-            future.addListener(new FutureListener<Object>() {
+            future.addListener(new FutureListener<Supplier<? extends SslContext>>() {
                 @Override
-                public void operationComplete(Future<Object> future) throws Exception {
+                public void operationComplete(Future<Supplier<? extends SslContext>> future) throws Exception {
                     try {
                         suppressRead = false;
                         if (future.isSuccess()) {
                             try {
-                                onSslContextOrSupplier(ctx, hostname, future.getNow());
+                                onSslContextSupplier(ctx, hostname, future.getNow());
                             } catch (Throwable cause) {
                                 ctx.fireExceptionCaught(new DecoderException(cause));
                             }
@@ -332,22 +333,70 @@ public class SniHandler extends ByteToMessageDecoder implements ChannelOutboundH
         }
     }
 
+    // Move these somewhere more accessible
+    private interface Function<X, Y> {
+        Y apply(X in);
+    }
+
+    private static <X, Y extends Future<X>, Z> GenericFutureListener<Y> cascadeTo(final Promise<Z> outPromise,
+                                                                                  final Function<X, Z> mapper) {
+        return new GenericFutureListener<Y>() {
+            @Override
+            public void operationComplete(Y future) throws Exception {
+                if (future.isSuccess()) {
+                    final Z result;
+                    try {
+                        result = mapper.apply(future.getNow());
+                    } catch (Throwable cause) {
+                        outPromise.setFailure(cause);
+                        return;
+                    }
+                    outPromise.setSuccess(result);
+                } else if (future.isCancelled()) {
+                    outPromise.cancel(false);
+                } else {
+                    outPromise.setFailure(future.cause());
+                }
+            }
+        };
+    }
+    // Move these somewhere more accessible
+
     /**
-     * The default implementation will simply call {@link SslContextMapping#map(String, Promise)} but users can
+     * The default implementation will simply call {@link SslContextMapping#map(Object, Promise)} but users can
      * override this method to implement custom behavior.
      *
      * NOTE: This method returns a {@code Future<?>} to accommodate the old and new way of doing things without
      * braking the API. Users who override this method are advised to change their return value from
      * {@code Future<SslContext>} to {@code Future<Supplier<? extends SslContext>>}.
      *
-     * @see SslContextMapping#map(String, Promise)
-     * @deprecated the return value will at some point change from {@code Future<SslContext>} to
-     *   {@code Future<Supplier<? extends SslContext>>}.
+     * @see SslContextMapping#map(Object, Promise)
+     * @deprecated Use {@link #lookupSupplier(io.netty.channel.ChannelHandlerContext, String)}.
      */
     @Deprecated
-    protected Future<?> lookup(ChannelHandlerContext ctx, String hostname) throws Exception {
-        Promise<Supplier<? extends SslContext>> promise = ctx.executor().newPromise();
-        return mapping.map(hostname, promise);
+    protected Future<SslContext> lookup(ChannelHandlerContext ctx, String hostname) throws Exception {
+        final Promise<SslContext> promise = ctx.executor().newPromise();
+        mapping.map(hostname, ctx.executor().<Supplier<? extends SslContext>>newPromise())
+                .addListener(cascadeTo(promise, new SniHandler.Function<Supplier<? extends SslContext>, SslContext>() {
+            @Override
+            public SslContext apply(Supplier<? extends SslContext> in) {
+                return in.get();
+            }
+        }));
+        return promise;
+    }
+
+    protected Future<Supplier<? extends SslContext>> lookupSupplier(ChannelHandlerContext ctx,
+                                                                    String hostname) throws Exception {
+        final Promise<Supplier<? extends SslContext>> promise = ctx.executor().newPromise();
+        lookup(ctx, hostname)
+                .addListener(cascadeTo(promise, new SniHandler.Function<SslContext, Supplier<? extends SslContext>>() {
+            @Override
+            public Supplier<? extends SslContext> apply(SslContext in) {
+                return new FixedSslContextSupplier(in);
+            }
+        }));
+        return promise;
     }
 
     /**
@@ -356,17 +405,8 @@ public class SniHandler extends ByteToMessageDecoder implements ChannelOutboundH
      * @see #select(ChannelHandlerContext, String)
      */
     @SuppressWarnings("unchecked")
-    private void onSslContextOrSupplier(ChannelHandlerContext ctx, String hostname, Object value) {
-        // NOTE: See JavaDoc of #lookup(...) to see why this if-else is necessary.
-        final Supplier<? extends SslContext> supplier;
-        if (value instanceof Supplier<?>) {
-            supplier = (Supplier<? extends SslContext>) value;
-        } else {
-            // If the user returned something other than a SslContext or Supplier<SslContext>
-            // in the Future<?> then let it blow up with a ClassCastException
-            supplier = new FixedSslContextSupplier((SslContext) value);
-        }
-
+    private void onSslContextSupplier(ChannelHandlerContext ctx, String hostname,
+                                      Supplier<? extends SslContext> supplier) {
         selection = new Selection(supplier, hostname);
         try {
             replaceHandler(ctx, hostname, supplier);

```